### ISSUE TITLE
Permit custom local maps for installed packages

### ISSUE THREAD

**Issue by guybedford** [jspm/jspm-cli#943]
Allow users to map dependencies of packages to any local mapping, and treating that as a manual override of the standard dependency resolution.

**Comment by Munter** [comment]
I just ran into this. For reasons I won't stand behind or argue for we have locally modified and checked in copies of third party libraries. I need to be able to configure that jquery or underscore should authoritatively be handled by these local files.

I've created an example repository with a SystemJS setup that behaves as I expect it to: https://github.com/Munter/jspm-local-dependency-bug/blob/master/app/config.js#L32

However, when I run `jspm install` it throws with
```
$ jspm install
     Looking up npm:babel-core
     Looking up npm:babel-runtime
     Looking up npm:core-js
     Looking up npm:backbone
     Updating registry cache...
ok   Up to date - babel as npm:babel-core@^5.6.4 (5.7.2)

err  TypeError: Arguments to path.resolve must be strings
         at Object.exports.resolve (path.js:313:15)
         at Package.getPath (/home/munter/.nvm/v0.10.29/lib/node_modules/jspm/lib/config/package-name.js:82:15)
         at Object.exports.download (/home/munter/.nvm/v0.10.29/lib/node_modules/jspm/lib/package.js:426:25)
         at Promise.resolve.then.then.resolution.fresh (/home/munter/.nvm/v0.10.29/lib/node_modules/jspm/lib/install.js:313:18)
         at lib$rsvp$$internal$$tryCatch (/home/munter/.nvm/v0.10.29/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:489:16)
         at lib$rsvp$$internal$$invokeCallback (/home/munter/.nvm/v0.10.29/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:501:17)
         at /home/munter/.nvm/v0.10.29/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:1095:13
         at Object.lib$rsvp$asap$$flush [as _onImmediate] (/home/munter/.nvm/v0.10.29/lib/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:1290:9)
         at processImmediate [as _immediateCallback] (timers.js:336:15)

warn Installation changes not saved.
``` 

I tracked it down to https://github.com/jspm/jspm-cli/blob/master/lib/package.js#L383 where it seems my locally handled dependency is now being downloaded, which fails because the local annotation doesn't have a registry.

So I tried to hack it:
```
exports.download = function(pkg, options, installDeps) {
  if (pkg.name.indexOf(':') === -1) {
    // Local package
    return Promise.resolve('local');
  }

  // ...
}
```

The above works-ish. jspm no longer throws when installing, the config is kept as I expect it to. The log is not so informative, but I bet someone who actually knows the code base would know how to fix that.

Would something like this be an approach? If you recognise a dependency to be configured as local, skip it and assume things work

**Comment by myitcv** [comment]
:+1: 

> For reasons I won't stand behind or argue for we have locally modified and checked in copies of third party libraries.

Not to take this thread off track but this may well be a totally reasonable approach in your circumstances. A very significant discussion happened on [`go-nuts`](https://groups.google.com/forum/#!forum/golang-nuts) on this very topic of vendoring.

**Comment by guybedford** [comment]
@Munter am i right in thinking that this workflow is working for you now with your change? Surely then you have the problem that an argumentless `jspm install` and other operations might clear the submap from your jspm config? I'd be happy to consider handling these edge cases and making sure that they can work out with manual changes.

In terms of the larger problem, I really think this is something peerDependencies should be able to fix, using an override when you install to set these items as peer dependencies, and using a local linking workflow to manage the packages. Basically peerDependencies would act like normal dependencies in jspm, but when there is an install conflict, users would be prompted to pick a resolution between them (like in Bower). A peerDependency would only need to be imposed by one parent package for it to expect everyone else to consolidate on a single option.

So imagine being able to install these dependencies with an override like:

```json
{
  "overrides": {
    "github:some/existing-package@1.2.3": {
      "peerDependencies": {
        "jquery": "local:jquery"
      }
    }
  }
}
```

Where `local:jquery` would be a paths configuration you set up, and a reserved namespace for any custom stuffs.

**Comment by guybedford** [comment]
Another point here - custom resolutions in the config file for say process to `npm:process` within React, don't seem to form part of the `jspm install` tree at all, which isn't ideal.

**Comment by guybedford** [comment]
Released in jspm 0.17 beta.
